### PR TITLE
Opens file in its default viewer. Has the same effect as double–click ing this file in Windows Explorer

### PR DESCRIPTION
**Pull request by Cyrillev91**
Add Option to Open file in its default viewer. (Has the same effect as double–clicking this file in Windows Explorer)

Example : you modify a file with notepad++ and you want to see this file in the default viewer 

- Without this feature : Right clic on tab then select "Open Containing Folder in Explorer" then double-clic on File
- With this feature : Right clic on tab then select "Open in Defaut Viewer"

With tab menu
openindefaultviewer

With File menu
openindefaultviewerfilemenu

Issue : #3576

### CODE REVIEW

File: PowerEditor/src/NppCommands.cpp [hunk 1]
```diff
@@ -105,6 +105,15 @@ void Notepad_plus::command(int id)
 		}
 		break;
 		
+		case IDM_FILE_OPEN_DEFAULT_VIEWER:
+		{
+			// Opens file in its default viewer. 
+            // Has the same effect as double–clicking this file in Windows Explorer.
+            BufferID buf = _pEditView->getCurrentBufferID();
+			ShellExecute(NULL, NULL, buf->getFullPathName(), NULL, NULL, SW_SHOW);
```
Comment (SinghRajenM) [inline PowerEditor/src/NppCommands.cpp]: What if there is no program associated with a particular file type. For example: one might have python script file (.py extension), then?

IMO, the "Open in Default Viewer" (under tab and File Menu) should be enabled only when a program/executable is associated with a particular file type.

Comment (Cyrillev91) [inline PowerEditor/src/NppCommands.cpp]: Hi,
If there is no association for a file type, widows will ask with it program to open it.
How do you open this file? (Comment voulez-vous ouvrir ce fichier ?)
windowsask

**Comment by SinghRajenM** [comment]
Correct. But the point is: there is no use of showing this dialog as you
will end up with opening file again either with npp or widows notepad (in
most of the cases).
So better enable this option only when associated program is available.
Consider using ```AssocQueryString``` API to check association.

**Comment by Cyrillev91** [comment]
I try this but is not perfect : 

        
	if (isFileExisting)
	{
		PTSTR ext = PathFindExtension(buf->getFullPathName());

		HRESULT hres;
		wchar_t buffer[MAX_PATH] = TEXT("");
		DWORD bufferLen = MAX_PATH;

		hres = AssocQueryString(NULL, ASSOCSTR_COMMAND, ext, NULL, buffer, &bufferLen);
				

		bool enable= (hres == S_OK)                                // check no no association in Windows XP or other error
			&& (wcsstr(buffer, TEXT("OpenWith.exe")) == NULL)  // check no no association in Windows 10
			&& (wcsstr(buffer, TEXT("OpenAs_RunDLL")) == NULL) // check no association in Windows 7
			&& (wcsstr(buffer, TEXT("notepad++")) == NULL);    // check association with notepad++

		_tabPopupMenu.enableItem(IDM_FILE_OPEN_DEFAULT_VIEWER, enable);
	}




1. **With file without association** : 

- With Windows XP : 
       - **hres** != S_OK 

- With Windows 7  : 
       - **hres** == S_OK 
       - **buffer** = C:\Windows\system32\rundll32.exe C:\Windows\system32\shell32.dll,OpenAs_RunDLL %1

- With Windows 10 : 
       - **hres** == S_OK 
       - **buffer** = C:\WINDOWS\system32\OpenWith.exe "%1" 

- Other Windows : no tested

- Next  Windows : ?

2. **With file associated with notepad++** : 
       - **hres** == S_OK 
       - **buffer** = C:\WINDOWS\system32\NOTEPAD.EXE %1




         
This is not perfect because one does not know the syntax in the next Windows, and the Exe of Notepad ++ could change name

What do you think ?

**Comment by SinghRajenM** [comment]
We are using 
```C++
hres = AssocQueryString(ASSOCF_VERIFY|ASSOCF_INIT_IGNOREUNKNOWN, ASSOCSTR_EXECUTABLE, ext, NULL, NULL, &bufferLen);
```
instead of 
```C++
hres = AssocQueryString(NULL, ASSOCSTR_COMMAND, ext, NULL, buffer, &bufferLen);
```

in our commercial products and have not faced any issue so far. In our case, we are not interested in app name, so we are passing ```NULL```.

> This is not perfect because one does not know the syntax in the next Windows, and the Exe of Notepad ++ could change name

If worried more, then consider using the return code of ```ShellExecute```. If return code<=32, then show error message to user similar to commit 7f98067c128a0fb00f3714c661c

**Comment by Cyrillev91** [comment]
Thank you for your help

**Pull request closed without merging**